File: example/mqtt_client.c
#include <mongoose.h>

static const char *s_address = "localhost:1883";
static const char *s_user_name = NULL;
static const char *s_password = NULL;
static const char *s_topic = "/stuff";
static struct mg_mqtt_topic_expression s_topic_expr = {NULL, 0};

static void ev_handler(struct mg_connection *nc, int ev, void *data)
{
	struct mg_mqtt_message *msg = (struct mg_mqtt_message *)data;

	switch (ev) {
		case MG_EV_CONNECT: {
			struct mg_send_mqtt_handshake_opts opts;
			memset(&opts, 0, sizeof(opts));
			opts.user_name = s_user_name;
			opts.password = s_password;

			mg_set_protocol_mqtt(nc);
			mg_send_mqtt_handshake_opt(nc, "dummy", opts);
			break;
		}

		case MG_EV_MQTT_CONNACK:
			if (msg->connack_ret_code != MG_EV_MQTT_CONNACK_ACCEPTED) {
				printf("Got mqtt connection error: %d\n", msg->connack_ret_code);
				exit(1);
			}
			s_topic_expr.topic = s_topic;
			printf("Subscribing to '%s'\n", s_topic);
			mg_mqtt_subscribe(nc, &s_topic_expr, 1, 42);
			break;

		case MG_EV_MQTT_PUBACK:
		  printf("Message publishing acknowledged (msg_id: %d)\n", msg->message_id);
	      break;
		
		case MG_EV_MQTT_SUBACK:
			printf("Subscription acknowledged, forwarding to '/test'\n");
			break;

		case MG_EV_MQTT_PUBLISH: {
#if 0
			char hex[1024] = {0};
			mg_hexdump(nc->recv_mbuf.buf, msg->payload.len, hex, sizeof(hex));
			printf("Got incoming message %.*s:\n%s", (int)msg->topic.len, msg->topic.p, hex);
#else
			printf("Got incoming message %.*s: %.*s\n", (int) msg->topic.len,
			msg->topic.p, (int) msg->payload.len, msg->payload.p);
#endif

			printf("Forwarding to /test\n");
			mg_mqtt_publish(nc, "/test", 65, MG_MQTT_QOS(0), msg->payload.p,
			msg->payload.len);
			break;
		}
		case MG_EV_CLOSE:
	      printf("Connection closed\n");
	      exit(1);
	}
}

static void signal_cb(struct ev_loop *loop, ev_signal *w, int revents)
{
	printf("Got signal: %d\n", w->signum);
	ev_break(loop, EVBREAK_ALL);
}

int main(int argc, char *argv[])
{
	struct ev_loop *loop = EV_DEFAULT;
	ev_signal sig_watcher;
	struct mg_mgr mgr;
	int i;

	ev_signal_init(&sig_watcher, signal_cb, SIGINT);
	ev_signal_start(loop, &sig_watcher);
	
	mg_mgr_init(&mgr, NULL);

	/* Parse command line arguments */
	for (i = 1; i < argc; i++) {
		if (strcmp(argv[i], "-h") == 0)
			s_address = argv[++i];
		else if (strcmp(argv[i], "-u") == 0)
			s_user_name = argv[++i];
		else if (strcmp(argv[i], "-t") == 0)
			s_topic = argv[++i];
		else if (strcmp(argv[i], "-p") == 0)
			s_password = argv[++i];
	}

	if (mg_connect(&mgr, s_address, ev_handler) == NULL) {
		fprintf(stderr, "mg_connect(%s) failed\n", s_address);
	}

	ev_run(loop, 0);

	printf("exit...\n");
	
	mg_mgr_free(&mgr);

	return 0;
}

File: example/lua/mqtt_client.lua
#!/usr/bin/lua

local ev = require("ev")
local evmg = require("evmongoose")
local loop = ev.Loop.default

local broker = arg[1] or "localhost:1883"
local keep_alive = 3

local alive_timer = ev.Timer.new(function(loop, w, event)
	keep_alive = keep_alive - 1
	if keep_alive == 0 then w:stop(loop) end
end, 5, 5)

local function ev_handle(con, event)
	if event == evmg.MG_EV_CONNECT then
		local result = con:get_evdata()
		if not result.connected then
			print("connect failed:", result.err)
			return
		end

		--[[
			Supported opt:
			user_name
			password
			client_id:		Default is: "evmongoose" .. mg_time()
			clean_session:	Default is false
			will_retain:	Default is false
		--]]
		
		local opt = {clean_session = true}
		con:mqtt_handshake(opt)
		
	elseif event == evmg.MG_EV_MQTT_CONNACK then
		local msg = con:get_evdata()
		
		if msg.code ~= evmg.MG_EV_MQTT_CONNACK_ACCEPTED then
			print("mqtt connection failed:", msg.err)
			return
		else
			print("mqtt connection ok")
		end

		local topic = {
			{name = "topic1"},
			{name = "topic2", qos = 1}
		}
		local mid = 12

		con:mqtt_subscribe(topic, mid)
		alive_timer:start(loop)

	elseif event == evmg.MG_EV_MQTT_SUBACK then
		local msg = con:get_evdata()
		print("conack mid:", msg.mid)
		
	elseif event == evmg.MG_EV_MQTT_PUBLISH then
		local msg = con:get_evdata()
		print("topic:", msg.topic)
		print("payload:", msg.payload)
		print("qos:", msg.qos)
		print("msgid:", msg.mid)

		con:mqtt_publish("test", "12345678")
		
	elseif event == evmg.MG_EV_MQTT_PINGRESP then
		print("Recv PingResp")
		keep_alive = 3

	elseif event == evmg.MG_EV_POLL then
		if keep_alive == 0 then
			-- Disconnect and reconnection
			con:set_flags(evmg.MG_F_CLOSE_IMMEDIATELY)
			keep_alive = 3
		end
	elseif event == evmg.MG_EV_CLOSE then
		print("connection close")
		
		ev.Timer.new(function()
			print("Try Reconnect to", broker)
			local mgr = con:get_mgr()
			mgr:connect(ev_handle, broker)
		end, 5):start(loop)
	end
end

local mgr = evmg.init(loop)
mgr:connect(ev_handle, broker)

ev.Signal.new(function(loop, sig, revents)
	loop:unloop()
end, ev.SIGINT):start(loop)

loop:loop()

print("exit...")
